### PR TITLE
[GSoC] Unify gestures with controls

### PR DESCRIPTION
**Pull request by BrayanDSO**
## Fixes
Closes #8078

## Approach
On the commits

## How Has This Been Tested?

Unit tests and on my device

https://user-images.githubusercontent.com/69634269/174163997-2e53d3ad-0ef3-465f-9469-13a5337d8e58.mp4

image

## Checklist
_Please, go through these checks before submitting the PR._

- [X] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [X] You have a descriptive commit message with a short title (first line, max 50 chars).
- [X] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [X] You have commented your code, particularly in hard-to-understand areas
- [X] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

**Comment by BrayanDSO** [comment]
Commits splitting only for aid reviewers. Please squash.

File: AnkiDroid/src/main/java/com/ichi2/anki/reviewer/Binding.kt [hunk 1]
```diff
@@ -253,7 +253,7 @@ class Binding private constructor(val modifierKeys: ModifierKeys?, val keycode:
         fun unknown(): Binding = Binding(ModifierKeys.none(), null, null, null)
 
         fun fromString(from: String): Binding {
-            if (TextUtils.isEmpty(from)) return unknown()
+            if (from.isEmpty()) return unknown()
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/reviewer/Binding.kt:256]: ❤️

**Review by david-allison — APPROVED** [review]
This looks great, thanks for cleaning out some of the gesture code

File: AnkiDroid/src/main/java/com/ichi2/preferences/ControlPreference.kt [hunk 2]
```diff
@@ -158,6 +172,41 @@ class ControlPreference : ListPreference {
...
+                bindings.remove(binding)
+                commandPreference.value = bindings.toPreferenceString()
+            }
+        }
+    }
+
+    private fun showDialogToReplaceBinding(binding: MappableBinding, title: String, parentDialog: MaterialDialog) {
+        val commandName = context.getString(getCommandWithBindingExceptThis(binding)!!.resourceId)
+
+        MaterialDialog.Builder(context)
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/preferences/ControlPreference.kt:222]: Just a note to have eyes open for @ShridharGoel with material dialog interaction #11646 (which already has conflicts and will likely take more - going to be tough to get that one in even though we are all motivated for it!)

**Review by mikehardy — APPROVED** [review]
Looks good to me, hope it has an impact on CI flakiness - would be a great result if so!

I am syncing translations prior to merging this one so we get the "real translation changes" out of the way in a separate PR prior to pushing this out since it is a bunch of i18n key motion

**Comment by mikehardy** [comment]
Can still flake - will still need attention I guess

```

UpgradeGesturesToControlsTest > [1: isValid(TestData(affectedPreferenceKey=gestureVolumeDown, keyCode=25, unaffectedPreferenceKey=gestureVolumeUp, binding=com.ichi2.anki.reviewer.MappableBinding@19753a))={1}] > com.ichi2.anki.servicemodel.UpgradeGesturesToControlsTest.invalid_preference_value_results_in_old_null_value_and_no_new_value[1: isValid(TestData(affectedPreferenceKey=gestureVolumeDown, keyCode=25, unaffectedPreferenceKey=gestureVolumeUp, binding=com.ichi2.anki.reviewer.MappableBinding@19753a))={1}] FAILED
    java.lang.AssertionError: Binding gestures should not be changed
    Expected: iterable containing ["preferenceUpgradeVersion", "gestureVolumeDown"]
         but: no item was "preferenceUpgradeVersion"
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
        at com.ichi2.anki.servicemodel.UpgradeGesturesToControlsTest.invalid_preference_value_results_in_old_null_value_and_no_new_value(UpgradeGesturesToControlsTest.kt:145)
UpgradeGesturesToControlsTest > [1: isValid(TestData(affectedPreferenceKey=gestureVolumeDown, keyCode=25, unaffectedPreferenceKey=gestureVolumeUp, binding=com.ichi2.anki.reviewer.MappableBinding@19753a))={1}] > com.ichi2.anki.servicemodel.UpgradeGesturesToControlsTest.if_mapped_to_non_empty_binding_then_added_to_end[1: isValid(TestData(affectedPreferenceKey=gestureVolumeDown, keyCode=25, unaffectedPreferenceKey=gestureVolumeUp, binding=com.ichi2.anki.reviewer.MappableBinding@19753a))={1}] FAILED
    java.lang.AssertionError: 
    Expected: iterable with items ["preferenceUpgradeVersion", "gestureVolumeDown", "binding_EDIT"] in any order
         but: no item matches: "preferenceUpgradeVersion", "gestureVolumeDown" in ["binding_EDIT"]
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:6)
        at com.ichi2.anki.servicemodel.UpgradeGesturesToControlsTest.if_mapped_to_non_empty_binding_then_added_to_end(UpgradeGesturesToControlsTest.kt:95)
```

File: AnkiDroid/src/test/java/com/ichi2/anki/servicemodel/UpgradeGesturesToControlsTest.kt [hunk 3]
```diff
@@ -37,64 +38,32 @@ import org.robolectric.ParameterizedRobolectricTestRunner
 import timber.log.Timber
 
 @RunWith(ParameterizedRobolectricTestRunner::class)
-class UpgradeVolumeButtonsToBindingsTest(private val testData: TestData) : RobolectricTest() {
+class UpgradeGesturesToControlsTest(private val testData: TestData) : RobolectricTest() {
```
[the remark this answers is not in the document]
Comment (BrayanDSO) [reply AnkiDroid/src/test/java/com/ichi2/anki/servicemodel/UpgradeGesturesToControlsTest.kt:41]: A lot. 

Poor tests feedback + no directions on how to fix them while taking quite some time to run them = headache

Separating the non parameterized tests will run them 50% less, so should help on them 🙏

**Pull request merged**